_posts/2016-02-24-publishing-data-with-odata-feeds.md

```diff
@@ -0,0 +1,100 @@
+So, OData is free and open source (Microsoft says...). In a sentence, it's a web protocol that allows publishing data over HTTP along with a description (**metadata**) of their structure and their relations to other entities.
+
+Why bother with structure and references to other entities? **SHOW ME THE DATA!** Well, metadata information about the data are used by clients like Excel to reconstitute data schema in their own mysterious ways and hydrate/map the data on the other side of the wire. Excel is a buzzword in our business so it didn't take long to put two and two together.
+
+## How did we do it...
+
+...as simple as possible. We used WebAPI 2 and OData version 4 to build endpoints which exposed entities/resources as IQueryable. So, we gathered all POCO classes used to store data in our mongo databases across the system and build simple readonly repositories to hydrate them from their data sources:
+```csharp
+public class SomeEntityDataStore : IEntityDataStore
+{
+  private readonly IMongoDatabase _mongoDatabase;
+
+  public SomeEntityDataStore(IMongoDatabase mongoDatabase)
+  {
+    _mongoDatabase = mongoDatabase;
+  }
+
+  public IQueryable<SomeEntity> GetAll()
+  {
+    return _mongoDatabase.GetCollection<SomeEntity>("someEntity").AsQueryable();   
+  }
+}
+```
+Connection strings for `mongoDatabase` were configured with `secondaryPreferred` option which basically channeled requests to secondary members in a mongodb server cluster if available. That way we take advantage of our replication strategy and off load OData read requests to the databases used less from our system.
+
+Repositories like that were injected into `ODataControllers`:
+```csharp
+public class SomeEntityController : ODataController
+{
+    private readonly IEntityDataStore _store;
+
+    public AuditorReportController(IEntityDataStore store)
+    {
+        _store = store;
+    }
+
+    [EnableQuery]
+    public IHttpActionResult Get()
+    {
+        return Ok(_store.GetAll());
+    }
+
+    [EnableQuery]
+    public IHttpActionResult Get([FromODataUri] string key)
+    {
+        return Ok(_store.GetAll().Where(entity => entity.Id == key));
+    }
+}
+```
+So, OData controllers supporting only HTTP GET requests expose the IQueryable source of some entity rendering them a readonly endpoint to that data source. An extra single result method was required to fetch entities by their id which is passed through the resource path. The `[EnableQuery]` marks the controller suitable to handle complex OData query expressions which are being passed from the client to the controller using querystring parameters. Thus, by exposing an IQueryable entity through an OData endpoint, query building becomes a responsibility of the client able to synthesize the OData query using expressions like `$filter`, `$select`, `$top`, `$skip` etc.
```

Comment:
> rendering them a readonly endpoint

rendering them as a readonly endpoint

----------

> marks the controller suitable to handle complex

marks the controller as suitable for handling the complex